src/PJ_deformation.c

```diff
@@ -0,0 +1,294 @@
+        dt = Q->t_obs - Q->t_epoch;
+    } else {
+        out = proj_coord_error(); /* in the 3D case +t_obs must be specified */
+        proj_log_debug(P, "deformation: +t_obs must be specified");
+        return out.xyz;
+    }
+
+    if (Q->has_xy_grids) {
+        shift = get_xygrid_shift(P, in.xyz);
+
+        out.xyz.x += dt * shift.x;
+        out.xyz.y += dt * shift.y;
+    }
+
+    if (Q->has_z_grids)
+        out.xyz.z += dt * get_zgrid_shift(P, in.xyz);
+
+    return out.xyz;
+}
+
+
+static PJ_OBS forward_obs(PJ_OBS in, PJ *P) {
+    struct pj_opaque *Q = (struct pj_opaque *) P->opaque;
+    double dt;
+    XYZ shift;
+    PJ_OBS out = in;
+
+    if (Q->t_obs != HUGE_VAL) {
+            dt = Q->t_obs - Q->t_epoch;
+        } else {
+            dt = in.coo.xyzt.t - Q->t_epoch;
+    }
+
+    if (Q->has_xy_grids) {
+        shift = get_xygrid_shift(P, in.coo.xyz);
+        out.coo.xyz.x += dt * shift.x;
+        out.coo.xyz.y += dt * shift.y;
+    }
+
+    if (Q->has_z_grids)
+        out.coo.xyz.z += dt * get_zgrid_shift(P, in.coo.xyz);
+
+    return out;
+}
+
+
+static LPZ reverse_3d(XYZ in, PJ *P) {
+    struct pj_opaque *Q = (struct pj_opaque *) P->opaque;
+    PJ_COORD out;
+    double dt = 0.0;
+    out.xyz = in;
+
+    if (Q->t_obs != HUGE_VAL) {
+        dt = Q->t_epoch - Q->t_obs;
+    } else {
+        out = proj_coord_error(); /* in the 3D case +t_obs must be specified */
+        proj_log_debug(P, "deformation: +t_obs must be specified");
+        return out.lpz;
+    }
+
+    if (Q->has_xy_grids) {
+        out.xyz = reverse_hshift(P, in, dt);
+    }
+
+    if (Q->has_z_grids)
+        out.xyz.z = in.z + dt * get_zgrid_shift(P, in);
+
+    return out.lpz;
+}
+
+static PJ_OBS reverse_obs(PJ_OBS in, PJ *P) {
+    struct pj_opaque *Q = (struct pj_opaque *) P->opaque;
+    PJ_OBS out = in;
+    double dt;
+
+    if (Q->t_obs != HUGE_VAL) {
+            dt = Q->t_epoch - Q->t_obs;
+        } else {
+            dt = Q->t_epoch - in.coo.xyzt.t;
+    }
+
+    if (Q->has_xy_grids)
+        out.coo.xyz = reverse_hshift(P, in.coo.xyz, dt);
+
+    if (Q->has_z_grids)
+        out.coo.xyz.z = in.coo.xyz.z + dt * get_zgrid_shift(P, in.coo.xyz);
+
+    return out;
+}
+
+static void *destructor(PJ *P, int errlev) {
+    if (0==P)
+        return 0;
+
+    if (0==P->opaque)
+        return pj_default_destructor (P, errlev);
+
+    if (P->opaque->cart)
+        P->opaque->cart->destructor (P->opaque->cart, errlev);
+
+    return pj_default_destructor(P, errlev);
+}
+
+
+PJ *PROJECTION(deformation) {
+    struct pj_opaque *Q = pj_calloc (1, sizeof (struct pj_opaque));
+    if (0==Q)
+        return pj_default_destructor(P, ENOMEM);
+    P->opaque = (void *) Q;
+
+    Q->cart = proj_create(P->ctx, "+proj=cart");
+    if (Q->cart == 0)
+        return destructor(P, ENOMEM);
+
+    /* copy ellipsoid parameters from P to Q->cart (simpler than guessing how */
+    /* the ellipsoid was specified in original definition)                    */
```

Comment:
Perhaps put a comment about this in PJ_cart.c, since this will bomb if algorithms are changed in PJ_cart in a way so additional ellipsoidal parameters are used.

Probably, this indicates the need for an internal `pj_inherit_ellipsoid_defs()`  function.

Reply:
I added the suggested function. No need to add further comments to PJ_cart.c.